agent/agent.go

```diff
@@ -313,20 +323,73 @@ type Agent struct {
 	// IP.
 	httpConnLimiter connlimit.Limiter
 
+	// Connection Pool
+	connPool *pool.ConnPool
+
 	// enterpriseAgent embeds fields that we only access in consul-enterprise builds
 	enterpriseAgent
 }
 
+type agentOption struct {
+	logger hclog.InterceptLogger
+	flags  *config.Flags
+}
+
+func WithLogger(logger hclog.InterceptLogger) agentOption {
+	return agentOption{
+		logger: logger,
+	}
+}
+
+func WithFlags(flags *config.Flags) agentOption {
+	return agentOption{
+		flags: flags,
+	}
+}
```

Comment:
The way I have generally seen this pattern implemented in Go is to use functions:

```go
type agentOption func(*agentOption)

func WithLogger(logger hclog.InterceptLogger) agentOption {
	return func(o *agentOption) {
		o.logger = logger
	}
}
```

Using functions we can delete `flattenAgentOptions`, which is likely to grow a lot over time as we add more options.

https://dave.cheney.net/2014/10/17/functional-options-for-friendly-apis

Reply:
Functions are fine too. I just took the pattern from our `testutil.WaitFor*` options but functions do make things a little easier. I did still have the flatten functions around but now they just loop over and execute all the functions to build out the final flattened options. So as more options are added no new code needs to be added to that function.